src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java

```diff
@@ -0,0 +1,328 @@
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.seqdoop.hadoop_bam.BAMInputFormat;
+import scala.Tuple2;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.*;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.Function;
+
+import static org.broadinstitute.hellbender.utils.Utils.calcMD5;
+
+@CommandLineProgramProperties(summary = "Runs BWA",
+        oneLineSummary = "BWA on Spark",
+        programGroup = SparkProgramGroup.class)
+public final class BwaSpark extends GATKSparkTool {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger log = LogManager.getLogger(BwaSpark.class);
+
+    private static ThreadLocal<Stopwatch> stopwatchThreadLocal = new ThreadLocal<Stopwatch>() {
+        @Override
+        protected Stopwatch initialValue() {
+            return Stopwatch.createUnstarted();
+        }
+    };
+
+    private static Set<Stopwatch> stopwatches = new HashSet<>();
+
+    @Argument(doc = "the reference", shortName = "ref",
+            fullName = "ref", optional = false)
+    private String ref;
+
+    @Argument(doc = "the output bam", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    private String output;
+
+    @Argument(doc = "the number of base pairs to send in a batch to BWA", shortName = "K",
+            fullName = "fixedChunkSize", optional = true)
+    private int fixedChunkSize = 100000;
+
+    @Argument(doc = "the number of threads", shortName = "t",
+            fullName = "threads", optional = true)
+    private int numThreads = 1;
+
+    @Override
+    public boolean requiresReads() {
+        return true;
+    }
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        Stopwatch toolStopwatch = Stopwatch.createStarted();
+        ctx.hadoopConfiguration().setBoolean(BAMInputFormat.KEEP_PAIRED_READS_TOGETHER_PROPERTY, true);
+        JavaRDD<GATKRead> unalignedReads = getReads();
+        JavaRDD<List<GATKRead>> unalignedPairs = unalignedReads.mapPartitions(iter -> () -> pairwise(iter));
+        JavaRDD<Tuple2<ShortRead, ShortRead>> shortReadPairs = unalignedPairs.map(p -> {
+            GATKRead read1 = p.get(0);
+            GATKRead read2 = p.get(1);
+            String name1 = read1.getName();
+            String name2 = read2.getName();
+            byte[] baseQualities1 = SAMUtils.phredToFastq(read1.getBaseQualities()).getBytes();
+            byte[] baseQualities2 = SAMUtils.phredToFastq(read2.getBaseQualities()).getBytes();
+            return new Tuple2<>(
+                    new ShortRead(name1, read1.getBases(), baseQualities1),
+                    new ShortRead(name2, read2.getBases(), baseQualities2));
+        });
+
+        JavaRDD<String> samLines = shortReadPairs.mapPartitions(iter -> () -> {
```

Comment:
can you extract everything up to `samLines` into one method and then the rest into another method? Ie, make it 3 clear steps: read, align, write